Overhaul CMake build

alexreinking opened the pull request:
This PR overhauls the CMake build to use modern CMake best practices. It bumps the minimum version to 3.15, which is available essentially everywhere. Ubuntu 20.04 LTS includes 3.16. Meanwhile, Visual Studio 2019 and Homebrew both provide _newer_ versions still.

The installed package honors a few different ways of switching the `tinyxml2::tinyxml2` alias between static and shared. If the user specifies `static` or `shared` as a component to `find_package`, it will use those (or fail). Otherwise, if `tinyxml2_SHARED_LIBS` is set, it will use that value to determine which to use (or fail). Otherwise, it will decay to standard CMake behavior, using `BUILD_SHARED_LIBS` as a hint. For more detail, [see my blog post on distributing both shared and static libs in CMake](https://alexreinking.com/blog/building-a-dual-shared-and-static-library-with-cmake.html).

I added GitHub Actions testing to build tinyxml2 as static/shared and debug/release. It runs the tests on all four configurations, both as part of the main build and through `find_package`, to ensure that it doesn't rot. Since the GHA testing covers both Travis and AppVeyor, I removed those files.

I removed support for Biicode, which has been dead for six years. Its website's SSL certificates have expired and some of the links on its GitHub repo now point to spam. Direct those users (if any remain) to vcpkg instead.

Fixes #804

---

If users want a single configuration of tinyxml2, it suffices to run:

```
$ cmake -G Ninja -S . -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=YES
$ cmake --build build
$ cmake --install build --prefix /path/to/install
```

Then their own build files don't need to be any more complicated than this:

```
cmake_minimum_required(VERSION 3.16)
project(example)

find_package(tinyxml2 REQUIRED)

add_executable(main main.cpp)
target_link_libraries(main PRIVATE tinyxml2::tinyxml2)
```

To build it, they would run:

```
$ cmake -G Ninja -S . -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=/path/to/install
$ cmake --build build
```

---

I think it would be good to evaluate whether continuing to support _seven (!)_ build systems is worthwhile. They aren't being tested rigorously by CI, so there's a looming threat of bit-rotting. CMake being a de-facto standard should put it in the running for being the primary build system. At the very least, if the other build systems are maintained by the community, the community should contribute GitHub Actions workflows.

For reference:

1. CMake
2. premake5
3. Meson ([compatible with CMake](https://mesonbuild.com/CMake-module.html))
4. Code::Blocks ([can be generated by CMake](https://cmake.org/cmake/help/latest/generator/CodeBlocks.html))
5. GNU Make ([can be generated by CMake](https://cmake.org/cmake/help/latest/generator/Unix%20Makefiles.html))
6. Xcode ([can be generated by CMake](https://cmake.org/cmake/help/latest/generator/Xcode.html))
7. Visual Studio ([can be generated by CMake](https://cmake.org/cmake/help/latest/generator/Visual%20Studio%2016%202019.html))

Because the source tree layout is so constrained by so many build systems, I had to awkwardly reference `xmltest.cpp` via directory traversal from `test/CMakeLists.txt`. Moving it (and `resources`) into the `test` folder would clean things up a bit, but I intentionally only touched CMake-related things in this PR.

Flamefire commented on CMakeLists.txt:
```diff
@@ -1,148 +1,126 @@
-
-project(tinyxml2)
-include(GNUInstallDirs)
+cmake_minimum_required(VERSION 3.16)
```
Do you **require** this recent version? E.g. some ubuntu distros are still at 3.10 so sticking to that makes adoption easier: https://packages.ubuntu.com/bionic/cmake
Yes I know updating CMake is easy, but I would require a recent version just because :)

alexreinking replied:
I didn't want to set a version earlier than I could easily test. I have 3.16 installed on Ubuntu 20.04 LTS. If you want to lower the version, I would welcome some assistance setting up GHA to use an older version.

alexreinking replied:
Certainly the version cannot be lowered below 3.14 (I think) as it stands because I use the newer features of GNUInstallDirs.

Flamefire replied:
On GHA you are already using an action which supports version selection: `lukka/get-cmake@v3.19.2` I think it makes sense, to use the one you specified as the minimum.
Looks like 3.15 is indeed required, didn't know about TARGET_FILE_BASE_NAME and would have just use what was there before. ;)

> newer features of GNUInstallDirs to automatically configure the install() commands

I'm interested: What are they?

alexreinking replied:
> > newer features of GNUInstallDirs to automatically configure the install() commands
> 
> I'm interested: What are they?

You can just write:

```
install(TARGETS foo EXPORT foo_exports)
```

Instead of:

```
install(
    TARGETS foo EXPORT foo_exports
    RUNTIME DESTINATION "${CMAKE_INSTALL_BINDIR}"
    LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}"
    ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
)
```

Basically it uses the correct GNUInstallDirs variables to set the destinations. It also introduced the `TYPE` argument: `TYPE INCLUDE` is the same as writing `DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"`. It's all just much cleaner. See the following for docs:

https://cmake.org/cmake/help/latest/release/3.14.html#commands

https://cmake.org/cmake/help/latest/command/install.html#installing-targets

https://cmake.org/cmake/help/latest/command/install.html#installing-files

alexreinking replied:
> On GHA you are already using an action which supports version selection: `lukka/get-cmake@v3.19.2` I think it makes sense, to use the one you specified as the minimum.

It does support version selection... but only as "low" as 3.17 🙂

Flamefire replied:
Thanks for the info, that was new to me :)
Oh, I see. Then maybe pip? Or https://github.com/marketplace/actions/actions-setup-cmake

Flamefire replied:
ah, this doesn't also install ninja. I see no need for that here, do you? Just omit the generator and let CMake choose

alexreinking replied:
I wanted to make sure that on Windows it wouldn't try to use Visual Studio (ie. a multi-config generator). I just dropped the min version to 3.15 and added testing for both the latest CMake and 3.15. Seems there's an issue with Windows on 3.15.

alexreinking commented:
@Flamefire - GitHub actions is now working. It tests on Windows (MSVC), Linux (GCC), and macOS (AppleClang). On each, it tests the minimum CMake version, 3.15, and the latest version (to catch potential backwards incompatibility issues).

alexreinking commented:
@Flamefire - just a gentle bump on this PR 🙂

Flamefire commented:
LGTM but I'm neither a maintainer nor a member here. Just checked the changes after your announcement on Slack, so my comments were just to help the maintainers on deciding :)

ajtruckle commented:
The own of this repository tends to do things in chunks. You will find that he will sign in and deal with a big bunch of commits and issues. He will handle this in time.

alexreinking commented:
> LGTM but I'm neither a maintainer nor a member here. Just checked the changes after your announcement on Slack, so my comments were just to help the maintainers on deciding :)

Ahhh, OK! I misunderstood your role.

Curve commented:
Will this get merged any time soon?
This CMake overhaul is badly needed

cugone commented:
The MacOS Github Action is not using an up-to-date compiler or the flags are incorrect. See PR #857 and its failing [MacOS build](https://github.com/leethomason/tinyxml2/pull/857/checks?check_run_id=2513170431) due to "scoped enum is a C++11 extension". I've opened a separate Issue #858 to track it.

alexreinking commented:
@cugone - I updated this PR to make sure that C++11 or newer is used by tinyxml2 and its linkees. However, this will not help with your PR unless this one is merged.

alexreinking commented:
As a first-time contributor, it seems I need @leethomason to approve running the GHA workflows. The workflows are passing on my fork: https://github.com/modern-cmake/tinyxml2/actions/runs/817601135

cugone commented:
Why did you delete the appveyor YAML file? It's required for Windows builds checks.

alexreinking commented:
> Why did you delete the appveyor YAML file? It's required for Windows builds checks.

From my top-level comment:

> I added GitHub Actions testing to build tinyxml2 as static/shared and debug/release. It runs the tests on all four configurations, both as part of the main build and through find_package, to ensure that it doesn't rot. **Since the GHA testing covers both Travis and AppVeyor, I removed those files.**

cugone commented:
> I updated this PR to make sure that C++11 or newer is used...

Now that I think about it, C++14 is a bug-fix for C++11 and that should be the minimum; but, I think if you're already going through the trouble of overhauling the entire build system anyway, updating to C++17 would be better.

Again, good luck with MacOS support. Getting that to compliant C++17 is a headache. At last check, they don't even support `std::filesystem` which has been out of experimental for every other compiler for at least 4 years.

alexreinking commented:
> C++14 is a bug-fix for C++11 and that should be the minimum; but, I think if you're already going through the trouble of overhauling the entire build system anyway, updating to C++17 would be better.

I'm not the maintainer of this project, so that's not my decision. I'm actually not sure now (in retrospect) that C++11 is meant to be a usage requirement for this project. Is this meant to be C++98 compatible? I would very much appreciate if the maintainer would weigh in here.

leethomason commented:
Woh! Time for a chunk update. I'll look at this in depth this weekend. Just scanning, there is no reason to have a C++ anything requirement, beyond C++ "classic" 98? The project doesn't use STL, and uses very vanilla templates.

I totally agree as well about the too many build systems points. I keep going back and forth. But there are 3 levels that make sense to me:
1. Remove the build systems entirely (except VS 2019 for convenience)
2. Use cmake + VS 2019 
3. Continue to leave what there is there

I took a hands off approach to the build, and I regret it. But I'm not sure the best path to fix it at this point.

leethomason commented:
Amending that thought. Approach #1 (only visual studio) breaks the valuable build testing on github. So (2) and (3) are the only real options.

cugone commented:
@leethomason  Just because a project doesn't use the STL or templates doesn't mean you shouldn't update to the modern C++ compilers. There are major advancements in the language behind the scenes that you are missing out on that will make the code more performant and better without changing anything or they can be applied incrementally via PRs; and, if you or PR do decide to use the newer features they'll be available.

alexreinking commented:
> Just because a project doesn't use the STL or templates doesn't mean you shouldn't update to the modern C++ compilers. There are major advancements in the language behind the scenes that you are missing out on that will make the code more performant and better without changing anything or they can be applied incrementally via PRs; and, if you or PR do decide to use the newer features they'll be available.
>> Woh! Time for a chunk update. I'll look at this in depth this weekend. Just scanning, there is no reason to have a C++ anything requirement, beyond C++ "classic" 98? The project doesn't use STL, and uses very vanilla templates.

Here's what I'll do: I'll comment out the line that adds a C++11 requirement to tinyxml2 so that nothing changes _now_, but it's easy for you to change your mind about it later.

cugone commented:
> Here's what I'll do: I'll comment out the line that adds a C++11 requirement to tinyxml2 so that nothing changes _now_, but it's easy for you to change your mind about it later. 

I would argue that's the exact opposite of what you should do. Leaving it as a C++11 requirement will allow modern techniques to be added without affecting current code. Yet, current code, with no changes, will get the benefits of advancements made due to the C++11 changes.

leethomason commented:
This PR does several things, which makes it a little challenging to disentangle. (Re-writes cmake, removes dead distribution systems, removes redundant project files, and there is even a C++ version discussion going on.) This PR is certainly leading to some important discussion. I've looked through this PR (and related) and gone back through recent PR history. HALF of all PRs against tinyxml2 is build / distribution related.

Which is weird, because TinyXML-2 is almost the most trivial thing you can build. One cpp, one h, and (for testing) one more cpp.

Clearly the build is an important and somewhat emotional discussion, but I'd like to get to the root of a couple questions. Because it's almost impossible for me to do the correct thing without understanding why this keeps coming up. I've been pleased to see several uses of TinyXML-2 "in the wild" and they've never used the build or distribution systems.
 
1. Why do we have a build system, at all? It's trivial to build and easily incorporated. Why so much work into the cmake (etc.) system? (And the automation could be done with simple: `gcc blah blah` command.)
2. Why the interest in distribution systems: vcpkg, etc? (I suspect this is related to #1 above, of course.)

My initial reaction was to simply delete the build and distribution systems. The source is cluttered by the overhead. But so much work continues to go into the build problem, that I really do want to understand the use case.

cugone commented:
> @cugone What improvements? That's completely hypothetical....

Not hypothetical. Here's [a non-exhaustive list of all the changes](https://en.cppreference.com/w/cpp/11) to the core language, library, and **741** bug fixes to the core language, and **868** bug fixes to the STL with the adoption of C++11.

> ...Which is weird, because TinyXML-2 is almost the most trivial thing you can build. One cpp, one h, and (for testing) one more cpp.

Not anymore. By lagging behind, newer compilers are starting to complain that you are using old and out-dated features. If you don't keep your library up-to-date and cling to using a broken standard that's been updated and fixed for 20 years, people stop using your library.

leethomason commented:
@cugone Regardless, we are conflating the C++ version discussion with what to do about the build & distribution. I'd like (for this PR) to focus on the build & distribution question. There is already too much to address in this PR.

leethomason commented:
@alexreinking I should also point out I like the change. The whole build thing has been due for clean up for a long time. I'm just trying to get a better understanding of the goals of the build system. 
There is a lot going in the PR. If you wanted to split it up, there's parts I'd take happily:
* Using cmake as the standard, deleting the other build systems (except VS because I use that and it's my project. :) )
* Moving everything to github actions and get rid of the the other CI tests

alexreinking commented:
> I should also point out I like the change. The whole build thing has been due for clean up for a long time. I'm just trying to get a better understanding of the goals of the build system.

Thanks. I should point out that the reason I did this work is because a colleague tried to use tinyxml2 through `FetchContent` (a CMake feature) and found that it did not work. I think your observation that your users "never used the build or distribution systems" is less because they aren't _interested_ in it than that they _just didn't work_. Sorry for my bluntness there, but I believe that's true.

> There is a lot going in the PR. If you wanted to split it up, there's parts I'd take happily:
>
>  * Using cmake as the standard, deleting the other build systems (except VS because I use that and it's my project. :) )
>  * Moving everything to github actions and get rid of the the other CI tests

That's great to hear. I will open a pair of PRs.

leethomason commented:
@alexreinking Don't worry about being blunt on whether the code is working for people - I'm genuinely trying to figure out the use case. (Also, you know, the pays the same either way. :) )

How did it not work? I mean, obviously he fetched it and it didn't compile, but was it for integration in another program as source? As a dll? Or just trying it out?

ldalessa commented:
> @alexreinking Don't worry about being blunt on whether the code is working for people - I'm genuinely trying to figure out the use case. (Also, you know, the pays the same either way. :) )
> 
> How did it not work? I mean, obviously he fetched it and it didn't compile, but was it for integration in another program as source? As a dll? Or just trying it out?

Hi @leethomason, that was probably me that @alexreinking referred to.

CMake's FetchContent has become one of the major tools with which C++ developers depend on small public github repositories exactly like this one. As a third-party user, I can simply add the following code to my CMakeLists.txt, and "magically" (through sufficiently advanced technology) depend on tinyxml2 as if it were provided by the system.

```
FetchContent_Declare(tinyxml2
  GIT_REPOSITORY     https://github.com/leethomason/tinyxml2.git
  GIT_TAG            master)
FetchContent_MakeAvailable(tinyxml2)
```

For better or worse, a decent fraction of the current C++ development community has moved to this pseudo-package style of external project dependencies. 

The CMake build infrastructure in tinyxml2 depends on ancient CMake functionality that is deprecated and will be removed, and triggers such warnings for our users. I believe @alexreinking has both modernized the functionality that was there in order to use current best-practices, and deduplicated build systems which become redundant given such modern CMake support.

The actual output for current master is.

```
CMake Deprecation Warning at gcc/_deps/tinyxml2-src/CMakeLists.txt:8 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


CMake Deprecation Warning at gcc/_deps/tinyxml2-src/CMakeLists.txt:9 (cmake_policy):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


CMake Deprecation Warning at gcc/_deps/tinyxml2-src/CMakeLists.txt:11 (cmake_policy):
  The OLD behavior for policy CMP0063 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


-- Configuring done
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   tinyxml2

```

leethomason commented:
@ldalessa and @alexreinking thanks for the context. I confess using cmake as a distribution system is new to me. (Not surprisingly, not being a cmake user to begin with.) Makes sense, and thanks for the explanation. @alexreinking don't worry about splitting the PR. I'll go ahead and test this.

leethomason commented:
@alexreinking Okay, looks good. Thanks for all the work on this! I'll be merging #861 which is this PR with some other project file wrapper fixes, and readme updates. 🤞

alexreinking commented:
Thanks @leethomason! Sorry I didn't have a chance to split it up like you'd asked.
